File: python/src/cmdop/resources/tunnels.py
"""tunnels namespace: open/close/list/get/sessions."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from cmdop._proto.cmdop.core.v1 import envelope_pb2 as pb
from cmdop._proto.cmdop.core.v1 import tunnels_pb2 as t_pb
from cmdop.resources.base import BaseResource

if TYPE_CHECKING:
    from collections.abc import AsyncIterator, Sequence

    from cmdop.types import ConnectedSessionOption, TunnelList, TunnelView


class TunnelsResource(BaseResource):
    async def open(
        self,
        session_id: str,
        *,
        local_port: int,
        protocol: str = "http",
        local_host: str = "127.0.0.1",
        subdomain: str | None = None,
        options: dict[str, str] | None = None,
    ) -> TunnelView:
        req = pb.Envelope(
            open_tunnel_req=t_pb.OpenTunnelRequest(
                session_id=session_id,
                protocol=protocol,
                local_host=local_host,
                local_port=local_port,
                subdomain=subdomain or "",
                options=options or {},
            )
        )
        return (await self._unary(req)).tunnel_view

    async def close(self, tunnel_id: str) -> None:
        """Close a tunnel (idempotent)."""
        req = pb.Envelope(close_tunnel_req=t_pb.CloseTunnelRequest(tunnel_id=str(tunnel_id)))
        await self._unary(req)

    async def list(self, *, page: int = 1, per_page: int | None = None) -> TunnelList:
        req = pb.Envelope(
            list_tunnels_req=t_pb.ListTunnelsRequest(page=self._offset_page(page, per_page))
        )
        return (await self._unary(req)).tunnel_list

    async def get(self, tunnel_id: str) -> TunnelView:
        req = pb.Envelope(get_tunnel_req=t_pb.GetTunnelRequest(tunnel_id=str(tunnel_id)))
        return (await self._unary(req)).tunnel_view

    async def sessions(self) -> Sequence[ConnectedSessionOption]:
        """Connected-session picker for opening tunnels."""
        req = pb.Envelope(list_sessions_req=t_pb.ListSessionsRequest())
        return list((await self._unary(req)).session_list.items)

    # -- pagination --------------------------------------------------------

    async def iter(self, *, per_page: int | None = None) -> AsyncIterator[Any]:
        async for page in self.pages(per_page=per_page):
            for item in page.items:
                yield item

    async def pages(self, *, per_page: int | None = None) -> AsyncIterator[TunnelList]:
        async def fetch(page_number: int) -> TunnelList:
            return await self.list(page=page_number, per_page=per_page)

        async for page in self._paginate_offset(fetch):
            yield page

File: node/src/resources/tunnels.ts
/**
 * `client.tunnels` — open / close / list / get / sessions.
 */

import type {
  ConnectedSessionOption,
  TunnelList,
  TunnelView,
} from "../_proto/cmdop/core/v1/tunnels_pb";
import { BaseResource, type ListPageOptions } from "./base";

export interface OpenTunnelOptions {
  protocol?: string;
  localHost?: string;
  localPort: number;
  subdomain?: string;
  options?: Record<string, string>;
}

export class TunnelsResource extends BaseResource {
  async open(sessionId: string, opts: OpenTunnelOptions): Promise<TunnelView> {
    const env = await this.unary(
      this.req({
        case: "openTunnelReq",
        value: {
          sessionId,
          protocol: opts.protocol ?? "http",
          localHost: opts.localHost ?? "127.0.0.1",
          localPort: opts.localPort,
          subdomain: opts.subdomain ?? "",
          options: opts.options ?? {},
        },
      }),
    );
    return env.payload.value as TunnelView;
  }

  /** Close a tunnel (idempotent). */
  async close(tunnelId: string): Promise<void> {
    await this.unary(this.req({ case: "closeTunnelReq", value: { tunnelId } }));
  }

  async list(opts: ListPageOptions = {}): Promise<TunnelList> {
    const env = await this.unary(
      this.req({ case: "listTunnelsReq", value: { page: this.offsetPage(opts.page, opts.perPage) } }),
    );
    return env.payload.value as TunnelList;
  }

  iter(opts: ListPageOptions = {}): AsyncGenerator<TunnelView> {
    return this.paginateOffset((page) => this.list({ page, perPage: opts.perPage }));
  }

  pages(opts: ListPageOptions = {}): AsyncGenerator<TunnelList> {
    return this.pagesOffset((page) => this.list({ page, perPage: opts.perPage }));
  }

  async get(tunnelId: string): Promise<TunnelView> {
    const env = await this.unary(this.req({ case: "getTunnelReq", value: { tunnelId } }));
    return env.payload.value as TunnelView;
  }

  /** Connected-session picker for opening tunnels. */
  async sessions(): Promise<ConnectedSessionOption[]> {
    const env = await this.unary(this.req({ case: "listSessionsReq", value: {} }));
    const list = env.payload.value as { items: ConnectedSessionOption[] };
    return list.items;
  }
}
